content/master/concepts/composition-functions.md

```diff
@@ -1,1026 +1,637 @@
+The `functions.yaml` file contains the Functions the Composition references in
+its pipeline steps:
 
-## Building a function
+```yaml
+---
+apiVersion: pkg.crossplane.io/v1beta1
+kind: Function
+metadata:
+  name: function-patch-and-transform
+spec:
+  package: xpkg.upbound.io/crossplane-contrib/function-patch-and-transform:v0.1.4
+```
+{{</expand>}}
 
- Crossplane doesn't have opinions about how a Composition Function is
- implemented. Functions must:
+`crossplane beta render` prints resources as YAML to stdout. It prints the
+composite resource first, followed by the resources the composition functions
+created.
 
- * Be packaged as an OCI image, where the `ENTRYPOINT` is the Function.
- * Accept input in the form of a `FunctionIO` document on stdin.
- * Return the `FunctionIO` they were passed, optionally mutated, on stdout.
- * Run within the constraints specified by the Composition that includes them,
-   such as timeouts, compute, network access.
+```yaml
+---
+apiVersion: example.crossplane.io/v1
+kind: XBucket
+metadata:
+  name: example-render
+---
+apiVersion: s3.aws.upbound.io/v1beta1
+kind: Bucket
+metadata:
+  annotations:
+    crossplane.io/composition-resource-name: storage-bucket
+  generateName: example-render-
+  labels:
+    crossplane.io/composite: example-render
+  ownerReferences:
+  - apiVersion: example.crossplane.io/v1
+    blockOwnerDeletion: true
+    controller: true
+    kind: XBucket
+    name: example-render
+    uid: ""
+spec:
+  forProvider:
+    region: us-east-2
+```
 
-This means Functions may be written using a general-purpose programming language
-like Python, Go, or TypeScript. They may also be implemented using a shell
-script, or an existing tool like Helm or Kustomize.
+{{<hint "important">}}
+Running `crossplane beta render` requires [Docker](https://www.docker.com).
 
-### FunctionIO
+See the [Crossplane CLI docs](https://github.com/crossplane/docs/pull/584) to
+learn how to install and use the Crossplane CLI.
+{{< /hint >}}
 
-When a Composition Function runner like `xfn` runs your Function it will write
-`FunctionIO` to its stdin. A `FunctionIO` is a Kubernetes style YAML manifest.
-It's not a custom resource (it never gets created in the API server) but it
-follows Kubernetes conventions.
+## Write a composition function
 
-A `FunctionIO` consists of:
+Composition functions let you replace complicated Compositions with code written
+in your programming language of choice. Crossplane has tools, software
+development kits (SDKs) and templates to help you write a function.
 
-* An optional, arbitrary `config` object.
-* The `observed` state of the XR, any existing composed resources, and their connection details.
-* The `desired` state of the XR and any composed resources.
-* Optional `results` of the Function pipeline.
 
-Here's a brief example of a `FunctionIO`:
+<!-- vale write-good.Passive = NO -->
+Here's an example of a tiny, hello world function. This example is written in
+Go.
```

Comment:
@plumbis, Vale seems to be parsing "Go." (with period) as a word and complaining that it's a spelling mistake. Any idea how to fix that? I added Go (no period) to the dictionary.